libcst/matchers/_matcher_base.py

```diff
@@ -103,6 +114,81 @@ def DoNotCare() -> DoNotCareSentinel:
     return DoNotCareSentinel.DEFAULT
 
 
+class TypeOf(Generic[_MatcherTypeT], BaseMatcherNode):
+    """
+    Matcher that matches any one of the given types. Useful when you want to work
+    with trees where a common property might belong to more than a single type.
+
+    For example, if you want either a binary operation or a boolean operation
+    where the left side has a name ``foo``::
+
+        m.TypeOf(m.BinaryOperation, m.BooleanOperation)(left = m.Name("foo"))
+
+    Or you could use the shorthand, like::
+
+        (m.BinaryOperation | m.BooleanOperation)(left = m.Name("foo"))
+
+    Also :class:`TypeOf` matchers can be used with initalizing in the default
+    state of other node matchers (without passing any extra patterns)::
+
+        m.Name | m.SimpleString
+
+    The will be equal to::
+
+        m.OneOf(m.Name(), m.SimpleString())
+    """
+
+    def __init__(self, *options: Union[_MatcherTypeT, "TypeOf[_MatcherTypeT]"]) -> None:
+        actual_options: List[_MatcherTypeT] = []
+        for option in options:
+            if isinstance(option, TypeOf):
+                if option.initalized:
+                    raise Exception(
+                        "Cannot chain an uninitalized TypeOf with an initalized one"
+                    )
+                actual_options.extend(option._raw_options)
+            else:
+                actual_options.append(option)
+
+        self._initalized = False
+        self._call_items: Tuple[Tuple[object, ...], Dict[str, object]] = ((), {})
+        self._raw_options: Tuple[_MatcherTypeT, ...] = tuple(actual_options)
+
+    @property
+    def initalized(self) -> bool:
+        return self._initalized
+
+    @property
+    def options(self) -> Iterator[BaseMatcherNode]:
+        for option in self._raw_options:
+            args, kwargs = self._call_items
+            matcher_pattern = option(*args, **kwargs)
+            yield matcher_pattern
+
+    def __call__(self, *args: object, **kwargs: object) -> BaseMatcherNode:
+        self._initalized = True
+        self._call_items = (args, kwargs)
+        return self
+
+    def __or__(
+        self, other: _OtherNodeMatcherTypeT
+    ) -> "TypeOf[Union[_MatcherTypeT, _OtherNodeMatcherTypeT]]":
+        return TypeOf[Union[_MatcherTypeT, _OtherNodeMatcherTypeT]](self, other)
+
+    def __and__(self, other: _OtherNodeMatcherTypeT) -> NoReturn:
+        left, right = type(self).__name__, other.__name__
+        raise TypeError(
+            f"TypeError: unsupported operand type(s) for &: {left!r} and {right!r}"
+        )
+
+    def __invert__(self) -> "AllOf[BaseMatcherNode]":
+        return AllOf(*map(DoesNotMatch, self.options))
+
+    def __repr__(self) -> str:
+        types = ", ".join(repr(option) for option in self._raw_options)
+        return f"TypeOf({types}, initalized = {self.initalized})"
```

Comment:
This is the only method not covered by unit tests. Would you like to also test it?

Reply:
If you have an aim about %100 coverage, sure, why not. But I dont usually test methods like `__repr__`.

Reply:
This is a minor issue and you can make the call.